src/video.c

```diff
@@ -129,8 +146,13 @@ void video_get_viewport_size(float *width, float *height) {
 
 void video_get_viewport(FloatRect *vp) {
 	video_get_viewport_size(&vp->w, &vp->h);
-	vp->x = (int)((video.current.width  - vp->w) / 2);
-	vp->y = (int)((video.current.height - vp->h) / 2);
+	if(video_is_highdpi()) {
+		vp->x = (int)(video.current.width  - (vp->w / 2));
+		vp->y = (int)(video.current.height - (vp->h / 2));
+	} else {
+		vp->x = (int)((video.current.width  - vp->w) / 2);
+		vp->y = (int)((video.current.height - vp->h) / 2);
+	}
```

Comment:
I have no idea what is going on in the true branch here, but I'm pretty sure it's wrong. First of all you can't use `video.current` here anymore, since the viewport size is calculated from `SDL_GL_GetDrawableSize` now, which will have a different scaling factor in high DPI mode.

The re-arranged formula in the true branch doesn't make sense either. This is supposed to implement letterboxing when the intended viewport's aspect ratio (constant 4:3 in our case) doesn't match that of the actual window. An offset is calculated such as to position the viewport in the center of the screen. This is done by subtracting the center of the viewport from the center of the window (and to find the center, you just divide the dimensions by two, hence the factored out `/2`).

If this actually looks ok on your system, I suspect it's only by sheer accident, and a different scaling factor will probably break it.